core/src/main/java/org/infinispan/interceptors/locking/PessimisticLockingInterceptor.java

```diff
@@ -134,54 +102,41 @@ public Object visitPrepareCommand(TxInvocationContext ctx, PrepareCommand comman
    }
 
    @Override
-   protected <K> Object handleWriteManyCommand(InvocationContext ctx, FlagAffectedCommand command, Collection<K> keys, boolean forwarded) throws Throwable {
-      Object maybeStage;
-      if (hasSkipLocking(command)) {
-         maybeStage = invokeNext(ctx, command);
+   protected <K> Object handleWriteManyCommand(InvocationContext ctx, FlagAffectedCommand command, Collection<K> keys,
+         boolean forwarded, int topologyId) throws Throwable {
+      final TxInvocationContext<?> txCtx = (TxInvocationContext<?>) ctx;
+      if (skipLockingForKeys(command, txCtx, keys)) {
+         txCtx.addAllAffectedKeys(keys);
```

Comment:
Related to https://github.com/infinispan/infinispan/pull/5442/files#r139728211 why do you add the keys here when these are not being locked?

Reply:
because the key is touched by the transaction.

Reply:
Yes, but previously the key was added to the affected set only if it was locked. Also, you test the locking (both primary and backup) through `context.getAffectedKeys().contains(key)` - so you need to decide; should this set contain only locked or all touched keys?

Reply:
touched keys. I didn't change the semantic of the set. See: https://github.com/infinispan/infinispan/blob/master/core/src/main/java/org/infinispan/interceptors/locking/PessimisticLockingInterceptor.java#L167

Reply:
OK, my bad that `handleWriteManyCommand` does not add them to the set correctly.
In that case, though, you need to change https://github.com/infinispan/infinispan/pull/5442/files#diff-922309dc5606a439d0cda2f8ab6ba470R186 and https://github.com/infinispan/infinispan/pull/5442/files#diff-922309dc5606a439d0cda2f8ab6ba470R190 because touched does not imply locked.

Reply:
touched keys implied they are copied to the transaction context. the key may change after being copied to the tx context. locking doesn't ensure nothing that's why we didn't do it before. 
https://github.com/infinispan/infinispan/blob/master/core/src/main/java/org/infinispan/interceptors/locking/PessimisticLockingInterceptor.java#L206

Reply:
What do you mean by "the key may change after being copied to the tx context"?
Locking does ensure that unless there's another operation with SKIP_LOCKING, this operation does atomically updates DC & persistence. I don't claim that the behaviour in master is correct.
Suppose that a transaction has been executed with one write using SL flag, and another write (on the same key) without the SL flag. The second operation must acquire the lock despite the key was touched before.

Reply:
using your example, the first write copies the key to the transaction context without locking the key. between the first and the second write, any transaction may commit. Acquire the locking in the second write doesn't giving you anything since you no longer have access to the most updated value. 

we said it in the javadoc:
```
So this flag is useful only as an optimization when the same key is written once and never again, or as an unsafe optimisation if the period between writes on the same key is large enough to make a race condition never happen in practice. If this is unclear, avoid it.
```

IMO, the behavior in master is correct

Reply:
@pruivo It gives you the guarantee that the cache won't end up with V1 in DC and V2 in cachestore, if there's a concurrent write writing V2. The javadoc also says "Bypasses lock acquisition for **this invocation**". You're right that you lose the correctness of the return value in the second write, but it's the same as if you execute a `get()` without forcing a write lock.
The period between the write w/SL and the second+concurrent write can be very long, so it matches even the suggestion later on.

Reply:
you're telling me what is written in the javadoc. the user is warned about the inconsistency between DC and CS. If he is using the flag, he knows what to expect. 
However, I still don't see why acquiring the locking in the second command would fix. Except if you wanna fix all the inconsistencies created by the SL flag. In that case, create a new JIRA :)

Reply:
BTW, forgot to say, if you want to remove the SKIP_LOCKING support from transactional caches, I'm 100% in favor :) SKIP_LOCKING and CACHE_MODE_LOCAL are the 2 flags that doesn't fit in the transactional  semantics

Reply:
@pruivo I don't care about inconsistencies due to SL in the first command - that's really expected. We don't try to fix those inconsistencies by design.
Acquiring the lock properly by the second command would guarantee that such transaction ends up with the cache in consistent state.
While ignoring SL in tx caches would be tempting, I think it would break the current implementation of atomic maps the least.

Reply:
@rvansa I agree that we use `SKIP_LOCKING` (and `CACHE_MODE_LOCAL`) with transactions too much internally in order to eliminate it now. Maybe if we had a way to do at least local writes without transactions in tx cache (like PFER, but more generic).

OTOH the `SKIP_LOCKING` javadoc does say it is for keys that do not change often, so a key being written to twice in the same transaction is clearly out of scope, and I wouldn't spend too much time trying to fix consistency in that case.

Reply:
@danberindei @pruivo As you wish, guys. We can fix it when we need it for internal works as we want to hide the flag anyway.
@pruivo May I ask you to document `getAffectedKeys()` that this is used for both *touched* and *locked* keys despite the key may not be actually locked?

Reply:
... and maybe a TODO in the interceptor. We've made a decision, let's document it so that we won't question it again after couple of months.